File: code/java/Exercises.java
import java.io.*;
import java.util.*;
import java.util.function.Predicate;
import java.util.stream.Stream;

// =========================================================
// Exercises class
// =========================================================
public final class Exercises {

    private Exercises() {}

    // 1) firstThenLowerCase
    public static Optional<String> firstThenLowerCase(List<String> xs, Predicate<String> predicate) {
        return xs.stream()
                 .filter(predicate)
                 .findFirst()
                 .map(s -> s.toLowerCase(Locale.ROOT));
    }

    // 2) say(...).and(...).phrase
    public static Say say() {
        return new Say("");
    }

    public static Say say(String word) {
        return new Say(word == null ? "" : word);
    }

    public static final class Say {
        private final String phrase;
        private Say(String phrase) { this.phrase = phrase; }

        public Say and(String word) {
            if (word == null) word = "";
            if (phrase.isEmpty()) return new Say(word);
            return new Say(phrase + " " + word);
        }

        public String phrase() { return phrase; }

        @Override public String toString() { return phrase; }
    }

    // 3) meaningfulLineCount
    public static long meaningfulLineCount(String filename) throws IOException {
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            return br.lines()
                     .map(String::stripLeading)
                     .filter(line -> !line.isEmpty())
                     .filter(line -> line.charAt(0) != '#')
                     .count();
        }
    }
}

// =========================================================
// Quaternion record
// =========================================================
record Quaternion(double a, double b, double c, double d) {

    public static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public static final Quaternion I    = new Quaternion(0, 1, 0, 0);
    public static final Quaternion J    = new Quaternion(0, 0, 1, 0);
    public static final Quaternion K    = new Quaternion(0, 0, 0, 1);

    public Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d))
            throw new IllegalArgumentException("Coefficients cannot be NaN");
    }

    public Quaternion plus(Quaternion q) {
        return new Quaternion(a + q.a, b + q.b, c + q.c, d + q.d);
    }

    public Quaternion times(Quaternion q) {
        double na = a*q.a - b*q.b - c*q.c - d*q.d;
        double nb = a*q.b + b*q.a + c*q.d - d*q.c;
        double nc = a*q.c - b*q.d + c*q.a + d*q.b;
        double nd = a*q.d + b*q.c - c*q.b + d*q.a;
        return new Quaternion(na, nb, nc, nd);
    }

    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }

    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }

    @Override public String toString() {
        StringBuilder sb = new StringBuilder();
        appendTerm(sb, a, "");
        appendTerm(sb, b, "i");
        appendTerm(sb, c, "j");
        appendTerm(sb, d, "k");
        return sb.length() == 0 ? "0" : sb.toString();
    }

    private void appendTerm(StringBuilder sb, double val, String suffix) {
        if (val == 0) return;
        if (sb.length() > 0) sb.append(val < 0 ? "" : "+");
        sb.append(val).append(suffix);
    }
}

// =========================================================
// BinarySearchTree sealed interface and implementations
// =========================================================
sealed interface BinarySearchTree permits Empty, Node {
    BinarySearchTree insert(String value);
    boolean contains(String value);
    int size();
}

final class Empty implements BinarySearchTree {
    public BinarySearchTree insert(String value) { return new Node(value, this, this); }
    public boolean contains(String value) { return false; }
    public int size() { return 0; }
    public String toString() { return ""; }
}

final class Node implements BinarySearchTree {
    private final String value;
    private final BinarySearchTree left;
    private final BinarySearchTree right;

    Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }

    public BinarySearchTree insert(String v) {
        int cmp = v.compareTo(value);
        if (cmp == 0) return this;
        if (cmp < 0) return new Node(value, left.insert(v), right);
        else return new Node(value, left, right.insert(v));
    }

    public boolean contains(String v) {
        int cmp = v.compareTo(value);
        if (cmp == 0) return true;
        return cmp < 0 ? left.contains(v) : right.contains(v);
    }

    public int size() {
        return 1 + left.size() + right.size();
    }

    public String toString() {
        return "(" + left + value + right + ")";
    }
}

File: code/kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.File
import java.io.IOException

fun <T> firstThenLowerCase(
    array: Array<T>
    predicate: (T) -> Boolean
): String? = array.firstOrNull(predicate)?.toString()?.lowercase()

class Say private constructor(private val words: List<String>) {
    val phrase: String
        get() = words.joinToString(" ")

    fun and (word: String): Say = 
        Say(words + word)

    companion object {
        fun start(first: String): Say =
            Say(listOf(first))
    }
}

fun countRelevantLines(filename: String): Int {
    return File(filename).bufferedReader().use { reader: BufferedReader ->
        reader.lineSequence()
            .filter { line ->
                val trimmed = line.trim()
                trimmed.isNotEmpty() && !trimmed.startsWith("#")
                }
                .count()
    }
}

data class Quaternion private constructor(
    val a: Double, 
    val b: Double, 
    val c: Double, 
    val d: Double 
) {
    operator fun plus(other: Quaternion): Quaternion = 
        Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)

    operator fun times(other: Quaternion): Quaternion {
        val e = other.a
        val f = other.b
        val g = other.c
        val h = other.d

        val ra = a * e - b * f - c * g - d * h
        val rb = a * f + b * e + c * h - d * g
        val rc = a * g - b * h + c * e + d * f 
        val rd = a * h + b * g - c * f + d * e 

        return Quaternion(ra, rb, rc, rd)
    }

    fun conjugate(): Quaternion = 
        Quaternion(a, -b, -c, -d)

    fun coefficients(): DoubleArray = 
        doubleArrayOf(a, b, c, d)

    override fun toString(): String = 
        "($a, $b, $c, $d)"

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)

        fun of(a: Double, b: Double, c: Double, d: Double): Quaternion = 
            Quaternion(a, b, c, d)
    }
}

sealed interface BinarySearchTree {
    val size: Int
    
    fun insert(value: String): BinarySearchTree

    fun contains(value: String): Boolean

    override fun toString(): String

    object Empty: BinarySearchTree {
        override val size: Int = 0
        
        override fun insert(value: String): BinarySearchTree = 
            Node(value, this, this)

        override fun contains(value: String): Boolean = false

        override fun toString(): String = ""
    }

    data class Node(
        val value: String,
        val left: BinarySearchTree,
        val right: BinarySearchTree
    ) : BinarySearchTree {
        override val size: Int = 1 + left.size + right.size

        override fun insert(value: String): BinarySearchTree = 
            when {
                value < this.value ->
                    Node(this.value, left.insert(value), right)
                value > this.value ->
                    Node(this.value, left, right.insert(value))
                else -> this
            }
        
        override fun contains(value: String): Boolean = 
            when {
                value < this.value -> left.contains(value)
                value > this.value -> right.contains(value)
                else -> true
            }
        
        override fun toString(): String = 
            "(${left.toString()}$value${right.toString()})"
    }
}